parser/docker_cmd.go

```diff
@@ -114,7 +120,151 @@ func (b *BuildCmd) Run() (string, error) {
 			"error":    nil,
 		},
 	})
-	return image.ID, nil
+
+	var retID = image.ID
+
+	if opts.Squash {
+		libsquash.Verbose = true
+
+		var imageReader1, pipeWriter1 = io.Pipe()
+		var imageReader2, pipeWriter2 = io.Pipe()
+		var squashedImageReader, squashedImageWriter = io.Pipe()
+
+		defer func() {
+			for _, pipeReader := range []*io.PipeReader{imageReader1, imageReader2, squashedImageReader} {
+				pipeReader.Close()
+			}
+
+			for _, pipeWriter := range []*io.PipeWriter{pipeWriter1, pipeWriter2, squashedImageWriter} {
+				pipeWriter.Close()
+			}
+		}()
+
+		// exporting in a goroutine since the io.Pipe is synchronous
+		go func() {
+			err := func() error {
+				b.reporter.Log(log.WithField("image_id", image.ID), "starting export of tar stream for squash")
+				b.reporter.Event(comm.EventOptions{
+					EventType: comm.BuildEventSquashStartExport,
+					Data: map[string]interface{}{
+						"image_id": image.ID,
+					},
+				})
+				exportOpts := docker.ExportImageOptions{
+					Name:         image.ID,
+					OutputStream: pipeWriter1,
+				}
+				if err := opts.DockerClient.Client().ExportImage(exportOpts); err != nil {
+					return err
+				}
+				b.reporter.Event(comm.EventOptions{
+					EventType: comm.BuildEventSquashFinishExport,
+					Data: map[string]interface{}{
+						"image_id": image.ID,
+					},
+				})
+				return nil
+			}()
+			if err != nil {
+				b.reporter.LogLevel(log.WithField("error", err), "error exporting image for squash", log.ErrorLevel)
+			}
+		}()
+
+		go func() {
+			err := func() error {
+				b.reporter.Log(log.WithField("image_id", image.ID), "starting export of tar stream for squash")
+				b.reporter.Event(comm.EventOptions{
+					EventType: comm.BuildEventSquashStartExport,
+					Data: map[string]interface{}{
+						"image_id": image.ID,
```

Comment:
The reason it needs two steams is because the files in the tarball are not
in any particular order. We need to go through the first stream in its
entirety to get all of the json files and to make our list of files that
should be in the squash layer's layer.tar. Then we go through the second to
get the actual files out of the individual layer.tars. As far as I can
tell, there's no way to duplex the stream without storing it in memory, so
using two allows us to throw away each when done with it.

On Sunday, January 4, 2015, Cole Gleason <notifications@github.com> wrote:

> In parser/docker_cmd.go
> <https://github.com/winchman/builder-core/pull/22#discussion-diff-22442372>
> :
>
> > +					},
> > +				})
> > +				return nil
> > +			}()
> > +			if err != nil {
> > +				b.reporter.LogLevel(log.WithField("error", err), "error exporting image for squash", log.ErrorLevel)
> > +			}
> > +		}()
> > +
> > +		go func() {
> > +			err := func() error {
> > +				b.reporter.Log(log.WithField("image_id", image.ID), "starting export of tar stream for squash")
> > +				b.reporter.Event(comm.EventOptions{
> > +					EventType: comm.BuildEventSquashStartExport,
> > +					Data: map[string]interface{}{
> > +						"image_id": image.ID,
>
> Why does libsquash needs tar streams of the same image twice?? Is this
> goroutine just the same as the above but with a different pipe?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/winchman/builder-core/pull/22/files#r22442372>.
>